### PR TITLE
fix: added check whether user is disabled on ldap server

### PR DESCRIPTION
**Pull request by voikin**
added a check whether the user is disabled on the ldap server using the `data` value in the error line

### CODE REVIEW

**Comment by ISulimanov** [comment]
@theburyat @yamaritta check plz

File: object/check.go [hunk 1]
```diff
@@ -284,9 +287,19 @@ func CheckLdapUserPassword(user *User, password string, lang string) (string, er
...
+			userDisabled = true
+			conn.Close()
+			break
+		}
+
 		conn.Close()
 	}
 
+	if userDisabled {
+		return "", fmt.Errorf("user is disabled")
```
Comment (yamaritta) [inline object/check.go:300]: А может ещё и в бд казгейт у пользователь стать признак заблокированнтго,чтобы мы таких могли фильтровать в ui например

File: object/check.go [hunk 2]
```diff
@@ -395,6 +408,10 @@ func CheckUserPassword(organization string, username string, password string, la
 				return nil, fmt.Errorf(i18n.Translate(lang, "check:The user: %s doesn't exist in LDAP server"), username)
 			}
 
+			if err.Error() == "user is disabled" {
+				return nil, fmt.Errorf(i18n.Translate(lang, "check:The user is forbidden to sign in, please contact the administrator"))
```
Comment (yamaritta) [inline object/check.go:412]: What do you think about user update in Casgate?
if we set a user as blocked, we can, for example, filter the list of users in UI

Comment (voikin) [reply object/check.go:412]: I think we can do it, it won't be a problem
we will just need to add a new field to the user table in db

Comment (ISulimanov) [reply object/check.go:412]: > I think we can do it, it won't be a problem we will just need to add a new field to the user table in db

field already exist user.IsForbidden

File: object/check.go [hunk 3]
```diff
@@ -34,6 +34,8 @@ const (
 
 	DefaultFailedSigninLimit      = 5
 	DefaultFailedSigninFrozenTime = 15
+
+	disabledLdapUserErrorData = "data 533"
```
Comment (ISulimanov) [inline object/check.go:38]: confused about how we check disabled ldap user,  @yamaritta Did you this approach before?

Comment (voikin) [reply object/check.go:38]: the problem is that if the user, for example, does not exist and if the user is blocked, then in both cases we will get errors that will differ by the `data` field in the error line

Comment (yamaritta) [reply object/check.go:38]: yes, we blocked user in  database when sync from ldap

Comment (yamaritta) [reply object/check.go:38]: but in our case ldap library helped us. if the current implementation does not allow get  account status, can leave the check as is

File: object/check.go (hunk 2 above)
Comment (voikin) [reply object/check.go:412]: i agree, this field is suitable

File: object/check.go (hunk 3 above)
Comment (ISulimanov) [reply object/check.go]: now lgfm, @yamaritta what about you?

File: object/check.go [hunk 4]
```diff
@@ -275,6 +276,17 @@ func CheckLdapUserPassword(user *User, password string, lang string) (string, er
 			return ldapServer.Id, fmt.Errorf(i18n.Translate(lang, "check:Multiple accounts with same uid, please check your ldap server"))
 		}
 
+		userDisabled = checkIsUserDisabled(searchResult.Entries[0].Attributes)
```
Comment (yamaritta) [inline object/check.go:279]: Are we sure that the first item is always exist?

File: object/check.go [hunk 5]
```diff
@@ -597,3 +617,15 @@ func CheckUserIdProviderOrigin(userIdProvider UserIdProvider) bool {
 
 	return isProviderNameEmpty != isLdapIdEmpty
 }
+
+func checkIsUserDisabled(userAttributes []*goldap.EntryAttribute) bool {
+	isDisabled := false
+
+	for _, attr := range userAttributes {
+		if attr.Name == "userAccountControl" && attr.Values[0] == "514" {
```
Comment (yamaritta) [inline object/check.go:625]: Are we sure that the first item is always exist?

File: object/check.go (hunk 3 above)
Comment (yamaritta) [reply object/check.go]: I remember userAccountControl is a bitmask. I'm not sure that this comparison is correct
https://learn.microsoft.com/ru-ru/troubleshoot/windows-server/active-directory/useraccountcontrol-manipulate-account-properties

File: object/check.go (hunk 4 above)
Comment (voikin) [reply object/check.go:279]:
```
                if len(searchResult.Entries) == 0 {
			conn.Close()
			continue
		}
		if len(searchResult.Entries) > 1 {
			conn.Close()
			return ldapServer.Id, fmt.Errorf(i18n.Translate(lang, "check:Multiple accounts with same uid, please check your ldap server"))
		}
```

above, here we check it out

File: object/check.go (hunk 3 above)
Comment (voikin) [reply object/check.go]: For a regular blocked account, the value of userAccountControl will always be 514:

`NORMAL_ACCOUNT (512) + ACCOUNTDISABLE (2) = 514`

File: object/check.go (hunk 5 above)
Comment (voikin) [reply object/check.go:625]: in lines 270-277 we check this

File: object/check.go (hunk 3 above)
Comment (yamaritta) [reply object/check.go]: Ok

**Review by yamaritta — APPROVED** [review]

**Pull request merged**